src/Termonad/App.hs

```diff
@@ -391,57 +394,74 @@ setupTermonad tmConfig app win builder = do
       else setShowTabs tmConfig note
 
   void $ onNotebookSwitchPage note $ \_ pageNum -> do
-    modifyMVar_ mvarTMState $ \tmState -> do
+    followUp <- modifyMVar mvarTMState $ \tmState -> do
       let notebook = tmStateNotebook tmState
           tabs = tmNotebookTabs notebook
           maybeNewTabs = updateFocusFL (fromIntegral pageNum) tabs
       case maybeNewTabs of
-        Nothing -> pure tmState
+        Nothing -> do
+          pure (tmState, pure ())
         Just (tab, newTabs) -> do
-          widgetGrabFocus $ tab ^. lensTMNotebookTabTerm . lensTerm
-          pure $
-            tmState &
-              lensTMStateNotebook . lensTMNotebookTabs .~ newTabs
+          let followUp = do
+                let newFocus = tab ^. lensTMNotebookTabFocusedTerm . lensTerm
+                widgetGrabFocus newFocus
```

Comment:
I'm glad you were able to figure out how to do this.  I should really document the need to pull out these types of follow-up IO actions and run them after modifying the `MVar`.  Hope this didn't take too long to figure out.

Reply:
It wasn't too bad; my initial attempt deadlocked, but I was able to figure out by adding a few print statements here and there.